src/approaches/grammar_search_invention_approach.py

```diff
@@ -432,93 +432,124 @@ def learn_from_offline_dataset(self, dataset: Dataset,
-                    yield (None, frozenset(s - {predicate}), 1.)
-
-            # Start the search with all of the candidates.
-            init = frozenset(candidates)
-        else:
-            assert CFG.grammar_search_direction == "smalltolarge"
-            # Successively consider larger predicate sets.
-            def _get_successors(s: FrozenSet[Predicate]
-                    ) -> Iterator[Tuple[None, FrozenSet[Predicate], float]]:
-                for predicate in sorted(set(candidates) - s):  # determinism
-                    # Actions not needed. Frozensets for hashing.
-                    # The cost of 1. is irrelevant because we're doing GBFS
-                    # and not A* (because we don't care about the path).
-                    yield (None, frozenset(s | {predicate}), 1.)
-
-            # Start the search with no candidates.
-            init = frozenset()
-
-        # Greedy best first search.
-        path, _ = utils.run_gbfs(
-            init, _check_goal, _get_successors, _heuristic,
-            max_evals=CFG.grammar_search_max_evals)
-        kept_predicates = path[-1]
-
-        print(f"Selected {len(kept_predicates)} predicates out of "
-              f"{len(candidates)} candidates:")
-        for pred in kept_predicates:
-            print(pred)
-
-        return set(kept_predicates)
+
+def _create_heuristic_function(initial_predicates: Set[Predicate],
+        atom_dataset: List[GroundAtomTrajectory],
+        candidates: Dict[Predicate, float]
+        ) -> Callable[[FrozenSet[Predicate]], float]:
+    if CFG.grammar_search_heuristic == "prediction_error":
+        return partial(_prediction_error_heuristic,
+                       initial_predicates,
+                       atom_dataset,
+                       candidates)
+    raise NotImplementedError(
+        f"Unknown heuristic: {CFG.grammar_search_heuristic}.")
+
+
+def _learn_operators_from_atom_dataset(atom_dataset: List[GroundAtomTrajectory]
+        ) -> Tuple[List[Segment], List[STRIPSOperator], List[OptionSpec]]:
+    """Relearn operators and option specs with the given dataset, which may
+    be pruned to include only a subset of the predicates..
+    """
+    segments = [seg for traj in atom_dataset
+                for seg in segment_trajectory(traj)]
+    strips_ops, partitions = learn_strips_operators(segments,
+                                                    verbose=False)
+    option_specs = [p.option_spec for p in partitions]
+    return (segments, strips_ops, option_specs)
+
+
+def _prediction_error_heuristic(initial_predicates: Set[Predicate],
+                                atom_dataset: List[GroundAtomTrajectory],
+                                candidates: Dict[Predicate, float],
+                                s: FrozenSet[Predicate]) -> float:
+    """Score a predicate set by learning operators and counting false positives.
+    """
+    print("Scoring predicates:", s)
```

Comment:
This seems a little strange to me. Aren't we redoing a lot of computation on every separate s, even if we don't need to? I think we should have these functions be called e.g. `_make_prediction_error_heuristic` which returns a Callable[_S, float].

Reply:
What do you mean? Of course we need to relearn operators for each new predicate set

Reply:
But we don't have to relearn on every new state! A heuristic-making function should return a function from state to float.

Reply:
Oh wait nvm I get it

Reply:
What? States are predicate sets here. Maybe I'm misunderstanding

Reply:
ya i got confused :D